src/util/alloc/immix_allocator.rs

```diff
@@ -126,48 +120,47 @@ impl<VM: VMBinding> Allocator<VM> for ImmixAllocator<VM> {
         offset: isize,
         need_poll: bool,
     ) -> Address {
-        trace!("{:?}: alloc_slow_once_precise_stress", self.tls);
+        trace!(
+            "{:?}: size {}, alloc_slow_once_precise_stress",
+            self.tls,
+            size
+        );
         // If we are required to make a poll, we call acquire_clean_block() which will acquire memory
         // from the space which includes a GC poll.
         if need_poll {
             trace!(
                 "{:?}: alloc_slow_once_precise_stress going to poll",
-                self.tls
+                self.tls,
             );
             let ret = self.acquire_clean_block(size, align, offset);
             // Set fake limits so later allocation will fail in the fastpath, and end up going to this
             // special slowpath.
             self.set_limit_for_stress();
             trace!(
                 "{:?}: alloc_slow_once_precise_stress done - forced stress poll",
-                self.tls
+                self.tls,
             );
             return ret;
         }
 
         // We are not yet required to do a stress GC. We will try to allocate from thread local buffer if possible.
-        // Restore the fake limit to the normal limit so we can do thread local alloaction normally.
+        // Restore the fake limit to the normal limit so we can do thread local allocation
+        // normally. Check if we have exhausted our current thread local block, and if so, then
+        // directly acquire a new one
         self.restore_limit_for_stress();
-        let ret = if self.alloc_slow_for_stress {
+        let ret = if self.require_new_block(size, align, offset) {
             // If we are already doing allow_slow for stress test, and reach here, it means we have failed the
             // thread local allocation, and we have to get a new block from the space.
             trace!(
                 "{:?}: alloc_slow_once_precise_stress - acquire new block",
-                self.tls
+                self.tls,
             );
             self.acquire_clean_block(size, align, offset)
         } else {
-            // Indicate that we are doing alloc slow for stress test. If the alloc() cannot allocate from
-            // thread local buffer, we will reach this method again. In that case, we will need to poll, rather
-            // than attempting to alloc() again.
-            self.alloc_slow_for_stress = true;
             // Try allocate. The allocator will try allocate from thread local buffer, if that fails, it will
             // get a clean block.
-            trace!("{:?}: alloc_slow_once_precise_stress - alloc()", self.tls);
-            let ret = self.alloc(size, align, offset);
-            // Indicate that we finish the alloc slow for stress test.
-            self.alloc_slow_for_stress = false;
-            ret
+            trace!("{:?}: alloc_slow_once_precise_stress - alloc()", self.tls,);
+            self.alloc(size, align, offset)
```

Comment:
Is this `alloc()` call guaranteed to succeed? If so, the comments above it need an update. If not, does that mean `alloc_slow_once_precise_stress()` is still reentrant (and we may have the issue of double counting some allocations)?

Reply:
Hmm. I am fairly certain it should succeed barring any OOM cases as the `require_new_block()` function basically acts like an `alloc()` call (note the hierarchy of checks in that function) without actually allocating anything. So if the function returns `false`, then we almost certainly have space for the current allocation request. Let me know if I should test this further.

Good catch regarding the comments. I will update them